### PR TITLE
[PivotTable] Add table cell

### LINKED ISSUE #797: [Pivot Table] PivotTableCell
# PivotTableCell

PivotTableCell is pivot table's cell, which contains the results of table's processing.

### **Visual example**

Screenshot from 2023-06-19 12-08-19

Hover behavior:
hover-behavior

From [figma](https://www.figma.com/file/TE9FUDtlgVQ4FWlAPtTagxQU/Bold-Design-System?type=design&node-id=33103-4697&t=FhcFEp346uCsjuul-0).

### PR DESCRIPTION
**Pull request by wesleydecezere**
closes #797

### CODE REVIEW

File: src/components/PivotTable/PivotTableCell/model.ts [hunk 1]
```diff
@@ -0,0 +1 @@
+export type PivotTableCellTypes = Array<'empty' | 'header' | 'value' | 'total' | 'grandtotal'>
```
Comment (wagnerbsantos) [inline src/components/PivotTable/PivotTableCell/model.ts:1]: Trocar de `Array` pra `Set`. Um `Set` é muito mais eficiente para as operações usadas.

**Review by wagnerbsantos — APPROVED** [review]
Ficou daora, só arruma o tipo que eu pedi ali

File: src/components/PivotTable/PivotTableCell/PivotTableCell.test.tsx [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+        const { backgroundColor: initialBgColor } = window.getComputedStyle(cell)
+        fireEvent.mouseEnter(cell)
+        fireEvent.mouseLeave(cell)
+        const { backgroundColor: finalBgColor } = window.getComputedStyle(cell)
+
+        expect(initialBgColor).toEqual(finalBgColor)
+      })
+    })
+
+    it("when cell types dont include 'value' or 'empty', shouldnt update bg color on mouse enter", () => {
```
Comment (zDudaHang) [inline src/components/PivotTable/PivotTableCell/PivotTableCell.test.tsx:77]:
```suggestion
    it("when cell types don't include 'value' or 'empty', shouldn't update background color on mouse enter", () => {
```

File: src/components/PivotTable/PivotTableCell/PivotTableCell.test.tsx [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+      const types = ['header'] as PivotTableCellTypes
```
Comment (zDudaHang) [inline src/components/PivotTable/PivotTableCell/PivotTableCell.test.tsx:78]:
```suggestion
      const typesWithHeader = ['header'] as PivotTableCellTypes
```
What do you think about this ?

File: src/components/PivotTable/PivotTableCell/PivotTableCell.test.tsx [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+      selectPivotTableCellElement(container, rowStart, columnStart)
+    )
+
+    expect(color).toEqual(expectedColor)
+    expect(backgroundColor).toEqual(expectedBackgroundColor)
+    expect(mockCalculateCellColor).toBeCalled()
+  })
+
+  describe('hover', () => {
+    describe.each(['value', 'empty'])("when cell type includes '%s'", (typeValue) => {
```
Comment (zDudaHang) [inline src/components/PivotTable/PivotTableCell/PivotTableCell.test.tsx:50]: I think is better to add the `typeValue` type

File: src/components/PivotTable/PivotTableCell/PivotTableCell.stories.tsx [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+import { PivotTableCellTypes } from './model'
+
+export default {
+  title: 'Components/PivotTable/PivotTableCell',
+}
+
+export const Default = () => {
+  const total = 9
+
+  const cellsTypeAndContent = [
```
Comment (zDudaHang) [inline src/components/PivotTable/PivotTableCell/PivotTableCell.stories.tsx:16]: Can you set the variable type ?

File: src/components/PivotTable/PivotTableCell/PivotTableCell.stories.tsx
Comment (zDudaHang) [inline src/components/PivotTable/PivotTableCell/PivotTableCell.stories.tsx:1]: I think this story must contain some Knobs like the [Button](https://github.com/laboratoriobridge/bold/blob/main/src/components/Button/Button.stories.tsx)

**Review by zDudaHang — CHANGES_REQUESTED** [review]
Great !

File: src/components/PivotTable/PivotTableCell/PivotTableCell.test.tsx (hunk 4 above)
Comment (wesleydecezere) [reply src/components/PivotTable/PivotTableCell/PivotTableCell.test.tsx:50]: The type casting still is necessary 😢 Maybe change this type to an array of enum maybe fix this

File: src/components/PivotTable/PivotTableCell/utils.test.ts [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+  })
+})
+
+describe('calculateCellColor', () => {
+  const theme = createTheme()
+  const defaultColor = theme.pallete.gray.c10
+  const defaultBgColor = theme.pallete.surface.main
+  const total = 10
+
+  it("when cell types dont include only 'value', should return default colors", () => {
```
[the remark this answers is not in the document]
Comment (wesleydecezere) [reply src/components/PivotTable/PivotTableCell/utils.test.ts:33]: The double quotes it's necessary to write `'value'` with this single quotes

File: src/components/PivotTable/PivotTableCell/styles.ts [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+  return {
+    root: {
+      gridArea: key,
+      backgroundColor: backgroundColor,
+      color: color,
+      borderTop: `1px solid ${theme.pallete.divider}`,
+      borderLeft: `1px solid ${theme.pallete.divider}`,
+      display: 'flex',
+      justifyContent: isValueOrEmpty ? 'flex-end' : 'flex-start',
+      fontWeight: (isHeaderOrTotal ? 'bold' : 'normal') as FontWeightProperty,
```
[the remark this answers is not in the document]
Comment (wesleydecezere) [reply src/components/PivotTable/PivotTableCell/styles.ts:28]: Yep. Honestly, I din't understand the reason for that

File: src/components/PivotTable/PivotTableCell/PivotTableCell.stories.tsx
Comment (wesleydecezere) [reply src/components/PivotTable/PivotTableCell/PivotTableCell.stories.tsx:1]: Nice, I will add this!

File: src/components/PivotTable/PivotTableCell/PivotTableCell.tsx [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+   */
+  isEndRow?: boolean
+  /**
+   * Additional styles
+   */
+  styles?: SerializedStyles
+}
+
+export function PivotTableCell(props: PivotTableCellProps) {
+  const { types, gridArea, children, isEndColumn, isEndRow, styles } = props
```
[the remark this answers is not in the document]
Comment (wesleydecezere) [reply src/components/PivotTable/PivotTableCell/PivotTableCell.tsx:39]: I didn't understand, do you think it's necessary to add default values for the nullable props?

File: src/components/PivotTable/PivotTableCell/PivotTableCell.tsx [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+  }
+
+  const handleMouseLeave = () => {
+    if (isValueOrEmpty) {
+      selectPivotTableCellElements(rowStart, columnStart).forEach((element) => element.removeAttribute('style'))
+    }
+  }
+
+  const rowNumbers = range(rowStart, rowEnd).join(' ')
+  const columnNumbers = range(columnStart, columnEnd).join(' ')
```
[the remark this answers is not in the document]
Comment (wesleydecezere) [reply src/components/PivotTable/PivotTableCell/PivotTableCell.tsx:81]: Yes, with `useMemo` it doesn't generate the sequence again when the boolean props change, for example.

File: src/components/PivotTable/PivotTableCell/classes/GridArea.test.ts [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+  })
+
+  it('when column end is greater than column start, should throw an error', () => {
+    const columnStart = 5
+    const columnEnd = 0
+
+    expect(() => new GridArea(0, columnStart, 0, columnEnd)).toThrowError(errorMessage)
+  })
+
+  it('should convert it to string correctly', () => {
```
Comment (zDudaHang) [inline src/components/PivotTable/PivotTableCell/classes/GridArea.test.ts:43]:
```suggestion
  it('should convert itself to string correctly', () => {
```

File: src/components/PivotTable/PivotTableCell/PivotTableCell.test.tsx [hunk 11]
```diff
@@ -0,0 +1,247 @@
...
+const createComponent = ({ maxValue, suffix }: PivotTableContextType, props: PivotTableCellProps) => {
+  return (
+    <PivotTableProvider value={{ maxValue, suffix }}>
+      <PivotTableCell {...props} />
+    </PivotTableProvider>
+  )
+}
+
+afterEach(() => {
+  jest.restoreAllMocks()
```
Comment (zDudaHang) [inline src/components/PivotTable/PivotTableCell/PivotTableCell.test.tsx:19]:
```suggestion
  jest.clearAllMocks()
```
No ?

File: src/components/PivotTable/PivotTableCell/PivotTableCell.stories.tsx [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+        {cellsTypeAndContent.map(({ type, content }, idx) => {
+          const types = new Set([type])
+          const gridArea = new GridArea(0, idx)
+
+          return (
+            <PivotTableCell
+              types={types}
+              key={gridArea.toString()}
+              gridArea={gridArea}
+              isEndRow={true}
```
Comment (zDudaHang) [inline src/components/PivotTable/PivotTableCell/PivotTableCell.stories.tsx:45]:
```suggestion
              isEndRow
```

**Review by zDudaHang — APPROVED** [review]
Great ! :rocket:

**Review by lucasjoao — APPROVED** [review]
Great! Just some details to check before the merge.

Do you plan to remove this component in pec codebase and after that start using this one there?

**Comment by wesleydecezere** [comment]
> Do you plan to remove this component in pec codebase and after that start using this one there?

Yes, but we plan to do this only after migrating all components of `PivotTable` from pec to here.

**Pull request merged**